broker/broker_test.go

```diff
@@ -0,0 +1,111 @@
+package broker
+
+import (
+	"sync"
+	"testing"
+)
+
+func TestNewSupervisor(t *testing.T) {
+	supervisor := NewSupervisor()
+	if supervisor.clientsMapping == nil {
+		t.Fatalf("error, expected clientsMapping to be a nil pointer, instead got %v", supervisor.clientsMapping)
+	}
+	if supervisor.clientsMutex == nil {
+		t.Fatalf("error, expected clientsMutex to be a nil pointer, instead got %v", supervisor.clientsMutex)
+	}
+	if supervisor.IsRunning != false {
+		t.Fatalf("error, expected IsRunning to be a false, instead got %v", supervisor.IsRunning)
+	}
+	if supervisor.clientOperationChannel == nil {
+		t.Fatalf("error, expected clientOperationChannel to be a nil pointer, instead got %v", supervisor.clientOperationChannel)
+	}
+	if supervisor.stoppedSignal == nil {
+		t.Fatalf("error, expected stoppedSignal to be a nil pointer, instead got %v", supervisor.stoppedSignal)
+	}
+	if supervisor.startedSignal == nil {
+		t.Fatalf("error, expected startedSignal to be a nil pointer, instead got %v", supervisor.startedSignal)
+	}
+	if supervisor.mainWg == nil {
+		t.Fatalf("error, expected mainWg to be a nil pointer, instead got %v", supervisor.mainWg)
+	}
+	if supervisor.clientCount != 0 {
+		t.Fatalf("error, expected clientCount to be a zero, instead got %v", supervisor.clientCount)
+	}
+}
+
+func TestRun(t *testing.T) {
+	var wg sync.WaitGroup
+	supervisor := NewSupervisor()
+	wg.Add(1)
+	go func() {
+		supervisor.START()
+		wg.Done()
+	}()
+	supervisor.WaitUntilReady()
+	if supervisor.IsRunning != true {
+		t.Fatalf("error, expected IsRunnint to be True, instead got %v", supervisor.IsRunning)
+	}
+	supervisor.STOP()
+	wg.Wait()
+	if supervisor.IsRunning != false {
+		t.Fatalf("error, expected IsRunnint to be False, instead got %v", supervisor.IsRunning)
+	}
+}
+
+func TestAddClient(t *testing.T) {
+	var wg sync.WaitGroup
+	supervisor := NewSupervisor()
+	wg.Add(1)
+	go func() {
+		supervisor.START()
+		wg.Done()
+	}()
+	supervisor.WaitUntilReady()
+	client := Client{
+		Id: "1",
+	}
+	supervisor.SignalClientAddition(&client)
+	supervisor.STOP()
+	wg.Wait()
+	clientCount := supervisor.GetClientCount()
```

Comment:
question: If we stopped the supervisor, the client counter maybe should be restarted ?

Reply:
As for now, we haven't decided yet which should be the full behavior of cases outside the happy path.
We need to meet and discuss this.